Handle stages and orchestrations that may reboot

jschmid1 opened the pull request:
This is a dummy implementation to tolerate reboots in stages.

This PR is dependent on https://github.com/SUSE/teuthology/pull/166

smithfarm commented on qa/tasks/deepsea.py:
```diff
@@ -742,14 +790,16 @@ def _run_stage_0(self):
+        if alternative_default:
+            # Dummy func call that adds the val of alt_def to the pillar
+            #               make that condition more flexible and less bad ..
+            if 'reboot' in alternative_default and not 'no-reboot' in alternative_default:
```
:-)

kshtsk commented on qa/tasks/deepsea.py:
```diff
@@ -720,6 +723,52 @@ def _run_orch(self, orch_tuple):
+        else:
+            self._exec(cmd_str, orch_spec)
+
+    def _exec_with_retry(self, cmd_str, orch_spec):
```
I would suggest refactor _exec method to support named argument `tries` defaults 1 instead of creating separate method.

jschmid1 replied:
The difference between `_exec_with_retry` and `_exec` is mainly the `ConnectionLostError` exception handling. As you can't construct a conditional in a try-except block, the refactoring to a single method would be a concatenation of two try-except blocks. Imo this looks bloated and doesn't read very nicely.

kshtsk replied:
Not sure I understand why it is bloated? My points:
1) Having two methods introduce code duplication
2) The methods actually do the same job especially one can use _exec_with_retry with max_attempts=1 without changes with the same result as _exec.

kshtsk replied:
@jschmid1 so, do I understand correctly that 
1) you want retry only if ConnectionLostError exception occurred?
2) you don't want catch ConnectionLostError in regular case.

jschmid1 replied:
Nope, I **only** want to check for ConnectionLostError when there is a reboot exepcted (check '_detect_reboot()')

kshtsk commented on qa/tasks/deepsea.py:
```diff
@@ -791,6 +809,57 @@ def _run_orch(self, orch_tuple):
+
+    def _exec_with_retry(self, cmd_str, orch_spec):
+        """
+        Execute cmd_str and catch CommandFiledError and
```
CommandFiledError => CommandFailedError

kshtsk commented on qa/tasks/deepsea.py:
```diff
@@ -791,6 +809,57 @@ def _run_orch(self, orch_tuple):
+        attempts = 0
+        sleep_time = 45
+        max_attempts = 5
+        while attempts < max_attempts:
```
@jschmid1 I actually meant this:
```
def _exec(self, cmd_str, orch_spec, max_attempts=1, sleep_time=45):
    """
    Execute cmd_str and catch CommandFiledError and
    ConnectionLostError and retry to run the command
    until one of the conditons are fulfilled:
    1) Execution succeeded
    2) Attempts are exceeded
    3) CommandFailedError is raised
    """
    attempts = max_attempts
    while attempts > 0:
        attempts -= 1
        try:
            self.master_remote.run(args=[
                'sudo', 'bash', '-c', cmd_str,
                ])
            break
        except CommandFailedError:
            log.error(
                "WWWW: orchestration {} failed. ".format(orch_spec)
                + "Here comes journalctl!"
                )
            self.master_remote.run(args=[
                'sudo',
                'journalctl',
                '--all',
                ])
            raise
        except ConnectionLostError:
            if attempts < 1:
                raise
            log.warning("No connection established yet..")
        log.info('{} from {} attempts left'.format(attempts, max_attempts))
        log.info("Sleeping for {} seconds".format(sleep_time))
        time.sleep(sleep_time)
```
This code will raise exception if you omit max_attempts (defaults=1) and will not raise any exception till the last try.

jschmid1 replied:
nice, I wasn't aware of that.

jschmid1 replied:
I'll re-implement that using the `safe_while` and your suggestion.

smithfarm commented:
@jschmid1 I just rebased ses6 branch on top of upstream `v14.0.1` - you'll need to rebase this PR (I do it by resetting to ses6 and then cherry-picking the commits). Sorry for the inconvenience!

While you're doing this, you might consider rebasing on top of `wip-qa-storage-roles` (PR #228) - that one is poised to go in.

jschmid1 commented:
@smithfarm done and done

smithfarm commented:
@jschmid1 You're a gentleman and a scholar!

smithfarm commented on qa/tasks/salt_manager.py:
```diff
@@ -227,6 +227,21 @@ def start_minions(self):
+    # stdout:    Minion did not return. [No response]
+    # (which means capturing the stdout and parsing it to ensure that
+    # the string "Minion did not return" did not appear after any of
+    # the minion names)
```
Please drop the comment block, above (it got pulled in inadvertently in your recent rebase, I assume).

smithfarm commented on qa/tasks/salt_manager.py:
```diff
@@ -227,6 +227,21 @@ def start_minions(self):
+
+    def enable_masters(self):
+        """Enables salt-master.service on all cluster nodes"""
+        self.__systemctl_cluster(subcommand="enable", service="salt-master")
```
Do we really want to enable salt-master.service on all nodes in the cluster?

smithfarm commented on qa/tasks/salt.py:
```diff
@@ -177,6 +177,8 @@ def setup(self):
         self.sm.start_master()
         self.sm.start_minions()
+        self.sm.enable_minions()
+        self.sm.enable_masters()
```
instead of `enable_masters` (plural) could we do `enable_master` (singular) like we have for `start_master`, above?

smithfarm commented on qa/tasks/deepsea.py:
```diff
@@ -734,6 +736,9 @@ def __init__(self, ctx, config):
         # cast stage/state_orch value to str because it might be a number
         self.stage = str(self.config.get("stage", ''))
         self.state_orch = str(self.config.get("state_orch", ''))
+        self.alternative_defaults = config.get('alternative_defaults', '')
```
The value of `alternative_defaults` is expected to be a list, right? In that case, would it make more sense for the default value here (second argument to `get()` method) to be `[]` instead of `''`?

smithfarm commented on qa/tasks/deepsea.py:
```diff
@@ -744,6 +749,19 @@ def __init__(self, ctx, config):
+        if self.alternative_defaults:
+            for cnf in self.alternative_defaults:
+                for k, v in cnf.items():
+                    info_msg = "Adjusting stage/orch to alternative-default {}: {}".format(k, v)
```
this line is too long - can you run this through flake8 ?

jschmid1 commented:
> @jschmid1 Please rebase. I just merged #228 (which includes the alternative_defaults code from this PR) into ses6.

I rebased atop ses6 and addressed the remaining suggestions. Please have a look again.

jschmid1 commented:
Oh, I will also get rid of my `_add_alternatives_to_pillar` for your `_maybe_apply_alternative_defaults`

smithfarm commented:
@jschmid1 Thanks - I slightly changed the nomenclature.

kshtsk commented:
@jschmid1 @smithfarm you guys love to put poems in function methods :-D

smithfarm commented:
> you guys love to put poems in function method names

@kshtsk It's much more prosaic than that, actually. See https://en.wikipedia.org/wiki/Self-documenting_code

kshtsk commented:
@smithfarm that feeling like you opened America to me ;-)

smithfarm commented:
@jschmid1 Nit: could you prefix the commit title with `qa/deepsea: ` like I just did for the PR title?

jschmid1 commented:
done @smithfarm

smithfarm commented on qa/tasks/deepsea.py:
```diff
@@ -851,36 +856,64 @@ def _run_orch(self, orch_tuple):
-            raise
-        self.log.info("orchestration {} completed".format(orch_spec))
+        tries = 0
+        if self.survive_reboots:
```
Can't we just use the survive reboots version for every orchestration?
In that case, we could dispense with the `survive_reboots` attribute, the `_detect_reboots` method, and the separate `_exec` method.

kshtsk replied:
I wonder if we do this always, then probably it is worth to push it to upstream.

jschmid1 replied:
> Can't we just use the survive reboots version for every orchestration?
> In that case, we could dispense with the `survive_reboots` attribute, the `_detect_reboots` method, and the separate `_exec` method.

I think we should clearly separate these cases. Imagine a case where machines reboot inadvertently. We would not be able to recognize that properly.

smithfarm replied:
OK, I agree, but in that case - in the spirit of "clear separation", I would prefer that `deepsea.orch` take an explicit "allow-reboot: true/false" config parameter, instead of relying on an (undocumented) heuristic that examines "alternative_default".

smithfarm replied:
Maybe leave the heuristic, in case someone forgets, but there should be a way to allow reboots without relying on `alternative_default`.

jschmid1 replied:
Agreed, pushed a version where allow_reboot takes presedence over the heuristic.

jschmid1 commented:
> If the new upgrade test case does not pass, it belongs in tier3, not tier2.

moving

smithfarm commented on qa/tasks/deepsea.py:
```diff
@@ -851,36 +856,73 @@ def _run_orch(self, orch_tuple):
+
+        if self.allow_reboot:
+            log.warning("Current orchestration may trigger a reboot")
+            return self.allow_reboot
```
nit: `return True` to match the rest of the code?

jschmid1 replied:
The idea behind is that you should be able to overwrite the heuristics by setting `allow_reboot` to false.

smithfarm replied:
If `allow_reboot` is set to false, then this return statement will never be hit.

smithfarm replied:
I guess you mean `if not self.allow_reboot:` and then `return False` (I really would prefer the explicit boolean constant in the `return`)

jschmid1 replied:
sigh.. right

jschmid1 replied:
Do we still want to be able to explicitly set allow_reboot to false?

smithfarm commented on qa/tasks/deepsea.py:
```diff
@@ -1,4 +1,5 @@
 import logging
+import time
```
`qa/tasks/deepsea.py:2:1: F401 'time' imported but unused`

jschmid1 replied:
good catch, fixing

smithfarm commented on qa/tasks/deepsea.py:
```diff
@@ -851,36 +856,73 @@ def _run_orch(self, orch_tuple):
+        2) Attempts are exceeded
+        3) CommandFailedError is raised
+        """
+        with safe_while(sleep=60, tries=tries, action="wait for reconnect") as proceed:
```
flake8 says this line is too long - please shorten

jschmid1 replied:
should we have automated flake8 checks in this repo if these rules are enforced?

smithfarm commented on qa/suites/deepsea/tier3/upgrade/tasks/upgrade.yaml:
```diff
@@ -0,0 +1,27 @@
+tasks:
+        - deepsea.preflight:
+        - deepsea.orch:
+                stage: prep
```
humor an old man and put "allow-reboot: true" here?

jschmid1 replied:
in

smithfarm commented on qa/suites/deepsea/tier3/upgrade/tasks/upgrade.yaml:
```diff
@@ -0,0 +1,27 @@
+                client.salt_master:
+                        - 'ceph -s'
+        - deepsea.orch:
+                state_orch: ceph.maintenance.upgrade
```
humor an old man and put "allow-reboot: true" here?

jschmid1 replied:
in

smithfarm commented on qa/tasks/deepsea.py:
```diff
@@ -851,36 +855,76 @@ def _run_orch(self, orch_tuple):
+        orchs_prone_to_reboot = ['ceph.maintenance.upgrade']
+
+        if self.allow_reboot is not None:
+            log.info("Setting allow_reboot explicitly to {}".
```
Please change `log.info` to `self.log.info` (the former does not set the prefix properly).

smithfarm commented on qa/tasks/deepsea.py:
```diff
@@ -851,36 +855,76 @@ def _run_orch(self, orch_tuple):
+        for cnf in self.alternative_defaults:
+            for k, v in cnf.items():
+                if 'reboot' in v and 'no-reboot' not in v:
+                    log.warning("Current orchestration may trigger a reboot")
```
Please change `log.info` to `self.log.info` (the former does not set the prefix properly).

smithfarm commented on qa/tasks/deepsea.py:
```diff
@@ -851,36 +855,76 @@ def _run_orch(self, orch_tuple):
+                    return True
+
+        if self.state_orch in orchs_prone_to_reboot:
+            log.warning("Current orchestration may trigger a reboot")
```
Please change `log.warning` to `self.log.warning` (the former does not set the prefix properly).

smithfarm commented on qa/tasks/deepsea.py:
```diff
@@ -851,36 +855,76 @@ def _run_orch(self, orch_tuple):
+                    break
+                except CommandFailedError:
+                    self.log.error(
+                        "WWWW: orchestration {} failed. ".format(orch_spec)
```
This should be:

```
            self.log.error(anchored(
                "orchestration {} failed. ".format(orch_spec)
```

(got reverted in your rebase)
